x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/inference/InferenceService.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.inference;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.inference.TaskType;
+import org.elasticsearch.xpack.core.inference.action.GetInferenceModelAction;
+import org.elasticsearch.xpack.core.inference.action.InferenceAction;
+import org.elasticsearch.xpack.esql.core.expression.FoldContext;
+import org.elasticsearch.xpack.esql.plan.logical.inference.InferencePlan;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+
+public class InferenceService {
+
+    private final Client client;
+
+    public InferenceService(Client client) {
+        this.client = client;
+    }
+
+    public ThreadContext getThreadContext() {
+        return client.threadPool().getThreadContext();
+    }
+
+    public void resolveInferences(List<InferencePlan> plans, ActionListener<InferenceResolution> listener) {
+
+        if (plans.isEmpty()) {
+            listener.onResponse(InferenceResolution.EMPTY);
+            return;
+        }
+
+        Set<String> inferenceIds = plans.stream()
+            .map(p -> p.inferenceId().fold(FoldContext.small()).toString())
+            .collect(Collectors.toSet());
+
+        CountDownLatch countDownLatch = new CountDownLatch(inferenceIds.size());
```

Comment:
We shouldn't use a sync `CountDownLatch` here. We can make it async with a `CountDownActionListener` or use the cluster state to retrieve the minimal service settings of the model. See https://github.com/elastic/elasticsearch/pull/121106 to access the model settings directly.